src/internal/pfsdb/commits.go

```diff
@@ -803,6 +799,100 @@ func ListCommit(ctx context.Context, db *pachsql.DB, filter CommitListFilter, re
+	var iter *commitIteratorTx
+	limit := 100
+	page, err := listCommitPage(ctx, tx, limit, 0, filter, reverse, revision)
+	if err != nil {
+		return nil, errors.Wrap(err, "list commit tx")
+	}
+	iter = &commitIteratorTx{
+		currRevision: -1,
+		revision:     revision,
+		reverse:      reverse,
+		commits:      page,
+		limit:        limit,
+		filter:       filter,
+		tx:           tx,
+	}
+	return iter, nil
+}
+
+func (iter *commitIteratorTx) Next(ctx context.Context, dst *CommitWithID) error {
+	if dst == nil {
+		return errors.Wrap(fmt.Errorf("commit is nil"), "get next commit")
+	}
+	var err error
+	if iter.index >= len(iter.commits) {
+		iter.index = 0
+		iter.offset += iter.limit
+		iter.commits, err = listCommitPage(ctx, iter.tx, iter.limit, iter.offset, iter.filter, iter.reverse, iter.revision)
+		if err != nil {
+			return errors.Wrap(err, "list commit page")
+		}
+		if len(iter.commits) == 0 {
+			return stream.EOS()
+		}
+	}
+	if iter.revision && iter.commits[iter.index].CreatedAt.After(iter.lastTimestamp) {
+		iter.currRevision++
+		iter.lastTimestamp = iter.commits[iter.index].CreatedAt
+	}
+	row := iter.commits[iter.index]
+	commit, err := getCommitInfoFromCommitRow(ctx, iter.tx, &row)
+	if err != nil {
+		return errors.Wrap(err, "getting commitInfo from commit row")
+	}
+	*dst = CommitWithID{
+		CommitInfo: commit,
+		ID:         row.ID,
+		Revision:   iter.currRevision,
+	}
+	iter.index++
+	return nil
+}
+
+func UpdateCommitTxByFilter(ctx context.Context, tx *pachsql.Tx, filter CommitListFilter, reverse, revision bool, cb func(commitWithID CommitWithID) error) error {
+	return errors.Wrap(listCommitTxByFilter(ctx, tx, filter, reverse, revision, func(commitWithID CommitWithID) error {
+		if err := cb(commitWithID); err != nil {
+			return err
+		}
+		return UpdateCommit(ctx, tx, commitWithID.ID, commitWithID.CommitInfo)
+	}), "list commits tx by filter")
+}
+
+func ListCommitTxByFilter(ctx context.Context, tx *pachsql.Tx, filter CommitListFilter, reverse, revision bool) ([]*pfs.CommitInfo, error) {
+	var commits []*pfs.CommitInfo
+	if err := listCommitTxByFilter(ctx, tx, filter, reverse, revision, func(commitWithID CommitWithID) error {
+		commits = append(commits, commitWithID.CommitInfo)
+		return nil
+	}); err != nil {
+		return nil, errors.Wrap(err, "list commits tx by filter")
+	}
+	return commits, nil
+}
+
+func listCommitTxByFilter(ctx context.Context, tx *pachsql.Tx, filter CommitListFilter, reverse, revision bool, cb func(commitWithID CommitWithID) error) error {
+	if len(filter) == 0 {
+		return errors.Errorf("filter cannot be empty")
+	}
+	for key, val := range filter {
+		if len(val) == 0 {
+			return errors.Errorf("filter values for key %q cannot be empty", key)
+		}
+	}
+	iter, err := listCommitTx(ctx, tx, filter, reverse, revision)
+	if err != nil {
+		return errors.Wrap(err, "create iterator for listCommitTx")
+	}
+	if err := stream.ForEach[CommitWithID](ctx, iter, func(commitWithID CommitWithID) error {
```

Comment:
Can we do `stream.Collect` and return the whole thing instead?

Reply:
Here we want to be able to iterate over the commits in the stream and potentially modify them (i.e. do an update) as we are iterating.